File: data_loader/data_loader.js
const pg = require('pg');
var pgtools = require("pgtools");
var fs = require('fs');
var program = require('commander');
const path = require('path');
const resultados = []
const databases = []
const promises = []
const promise_database = []

// Definir parámetros y descripción del contenido del parámetro.
program
  .argument('<file_json>', 'Input file in json format with the data to be uploaded to the database')
  .argument('<database>', 'Database name to be used to upload the data, with the option -c or --create is created it')
  .argument('<user>', 'Database user, this user is specified in the config file docker-compose.yml')
  .argument('<password>', 'Database password for the user specified in the config file docker-compose.yml')
  .argument('<host>', 'Database host (default: localhost)')
  .option('-c, --create', 'with this option the database is created if it does not exist, this argument is optional')
  .argument('<table>', 'Table name to be used to upload the data, if it does not exist it is created')

program.parse(process.argv);

const options = program.args.concat(program.opts())


const connectionData = {
  database: options[1],
  user: options[2],
  password: options[3],
  host: options[4],
}
const config_create = {
  user: options[2],
  password: options[3],
  host: options[4]
};


const cliente = new pg.Client(connectionData);
const cliente_create = new pg.Client(config_create);

async function createDatabase(database) {
  return new Promise((resolve, reject) => {
    pgtools.createdb(config_create, database, (err) => {
      if (err) {
        console.log("¡The database is going to be deleted and created again.!")
        resolve()
      }
      else {
        console.log('Database created')
        resolve()
      }
    })
  })
}

async function deleteDatabase(database) {
  return new Promise((resolve, reject) => {
    pgtools.dropdb(config_create, database, (err) => {
      if (err) {
        console.log("Error deleting database\n\n" + err)
        process.exit()
      } else {
        console.log('Database deleted')
        resolve()
      }
    })
  })
}

async function checkDatabase(database) {
  cliente.connect(err => {
    const query_database = `SELECT datname FROM pg_catalog.pg_database;`
    promise_database.push(cliente.query(query_database).then(result => {
      result.rows.map(row => {
        databases.push(row.datname)
      })
    }).catch(() => {
      if (!databases.includes(database)) {
        console.log('ERROR: Database with this properties not exists')
        process.exit()
      }
    }))
  })
}

async function createTable(file) {
  const query_table = `SELECT table_name FROM information_schema.tables WHERE table_type = 'BASE TABLE' AND table_schema = 'public'`;
  cliente.query(query_table).then(result => {
    result.rows.map(row => {
      resultados.push(row.table_name)
    })
    if (!resultados.includes(options[5])) {
      console.log('Creating table...')
      const query_create = `CREATE TABLE ${options[5]}(id SERIAL, time DOUBLE PRECISION, departureTime TIMESTAMP, arrivalTime TIMESTAMP,dayOfWeek VARCHAR(10), plate VARCHAR(10), time_short TIME, prediction DOUBLE PRECISION)`;
      cliente.query(query_create)
        .catch(err => {
          console.log(err)
        })
    }
    dataUploadToTimeScale(file)
  })
}

async function dataUploadToTimeScale(file) {
  FILE_NAME = file;
  let filename = fs.readFileSync(FILE_NAME)
  try {
    let parse_json = JSON.parse(filename)
    sort(parse_json, 'time')



    //console.log(parse_json)
    console.log('Inserting data into database...')
    for (var item in parse_json) {
      var time_short = parse_json[item].arrivalTime.split('T')[1].split(':')
      if (time_short[2] == undefined) {
        time_short[2] = '00'
      }
      var time_short_string = time_short[0] + ':' + time_short[1] + ':' + time_short[2]
      var hours = (Date.parse(parse_json[item].departureTime) - Date.parse(parse_json[item].arrivalTime)) / (1000)
      const query = `INSERT INTO ${options[5]}(id, time, departureTime, arrivalTime, dayOfWeek, plate, time_short) VALUES(DEFAULT, ${hours}, '${(parse_json[item].departureTime).replace("T", " ")}', '${(parse_json[item].arrivalTime).replace("T", " ")}', '${parse_json[item].dayOfWeek}', '${parse_json[item].plate}', '${time_short_string}')`;
      promises.push(cliente.query(query).then(() => {
      }))
    }
    Promise.all(promises).then(() => {
      process.exit()
    })
  }
  catch {
    console.log('ERROR in input file. It is not in JSON format.')
    process.exit()
  }
}

function sort(json, key) {
  json.sort(function (a, b) {
    return a[key] > b[key];
  });
}

if (options.length === 7) {
  if (options[6].create) {
    async function create() {
      console.log('Creating database...')
      await createDatabase(options[1])
      if (createDatabase[options[1]] === undefined) {
        await deleteDatabase(options[1])
        await createDatabase(options[1])
      }
      await checkDatabase(options[1])
      await createTable(options[0])
    }
    if (fs.existsSync(options[0])) {
      cliente_create.connect(err => {
        if (err) {
          if (err.message.includes('password authentication failed')) {
            console.log('ERROR: The user or password is wrong')
            process.exit()
          }
          if (err.code === 'ENOTFOUND') {
            console.log('ERROR: The host is wrong')
            process.exit()
          }
        }
        else {
          create()
        }
      })
    } else {
      console.log('ERROR: Input file ' + options[0] + ' is not a file')
      process.exit()
    }
  } else {
    if (fs.existsSync(options[0])) {
      checkDatabase(options[1])
      cliente.connect(err => {
        if (err) {
          if (err.message.includes('password authentication failed')) {
            console.log('ERROR: The user or password is wrong')
            process.exit()
          }
          if (err.code === 'ENOTFOUND') {
            console.log('ERROR: The host is wrong')
            process.exit()
          }
        }
        else {
          Promise.all(promise_database).then(() => {
            createTable(options[0])
          })
        }
      })
    } else {
      console.log('ERROR: Input file ' + options[0] + ' is not a file')
      process.exit()
    }
  }
}
else {
  console.log('Usage: node data-upload.js <-c> <file.json> <database> <user> <password> <localhost> <table>\n\n<-c> optional argument for create database \n <file.json>: Name of the json format file with the data to load in TimeScale \n<database>: PostgreSQL database name \n<user>: User associated with the database \n<password>: Database password \n<localhost>: Host\n<table>: Table name in the database or name of the table to create')
}

File: data_loader/data_loader.py
import json
import argparse
import sys
import os
from datetime import datetime
import psycopg2
from psycopg2 import sql
from psycopg2.extras import execute_batch

text = "Load data from JSON file into TimescaleDB database"

parser = argparse.ArgumentParser(description=text)
parser._action_groups.pop()
required = parser.add_argument_group('required arguments')
optional = parser.add_argument_group('optional arguments')

required.add_argument("file_json", help="Input file in json format with the data to be uploaded to the database")
required.add_argument("database", help="Database name to be used to upload the data, with the option -c or --create is created it")
required.add_argument("user", help="Database user, this user is specified in the config file docker-compose.yml")
required.add_argument("password", help="Database password for the user specified in the config file docker-compose.yml")
required.add_argument("host", help="Database host (default: localhost)")
required.add_argument("table", help="Table name to be used to upload the data, if it does not exist it is created")
optional.add_argument("-c", "--create", action='store_true', help="with this option the database is created if it does not exist, this argument is optional")

args = parser.parse_args()

def create_connection(dbname, user, password, host, autocommit=False):
    """Create a PostgreSQL connection"""
    try:
        conn = psycopg2.connect(
            dbname=dbname,
            user=user,
            password=password,
            host=host
        )
        if autocommit:
            # Needed for operations like CREATE DATABASE that cannot run inside a transaction block
            conn.autocommit = True
        return conn
    except psycopg2.Error as e:
        print(f"Unable to connect to database '{dbname}': {e}")
        sys.exit(1)

def check_database_exists(conn, database_name):
    """Check if a database exists"""
    try:
        cursor = conn.cursor()
        cursor.execute("SELECT datname FROM pg_catalog.pg_database WHERE datname = %s", (database_name,))
        return cursor.fetchone() is not None
    except psycopg2.Error as e:
        print(f"Error checking database existence: {e}")
        return False

def create_database(conn, database_name):
    """Create a database"""
    try:
        conn.autocommit = True
        cursor = conn.cursor()
        cursor.execute(sql.SQL("CREATE DATABASE {}").format(sql.Identifier(database_name)))
        print(f"Database '{database_name}' created")
        return True
    except psycopg2.Error as e:
        if "already exists" in str(e):
            print(f"Database '{database_name}' already exists")
            return True
        else:
            print(f"Error creating database: {e}")
            return False

def check_table_exists(conn, table_name):
    """Check if a table exists"""
    try:
        cursor = conn.cursor()
        cursor.execute("""
            SELECT table_name 
            FROM information_schema.tables 
            WHERE table_type = 'BASE TABLE' 
            AND table_schema = 'public'
            AND table_name = %s
        """, (table_name,))
        return cursor.fetchone() is not None
    except psycopg2.Error as e:
        print(f"Error checking table existence: {e}")
        return False

def create_table(conn, table_name):
    """Create the trucks table"""
    try:
        cursor = conn.cursor()
        create_table_query = sql.SQL("""
            CREATE TABLE {} (
                id SERIAL PRIMARY KEY,
                time DOUBLE PRECISION,
                departureTime TIMESTAMP,
                arrivalTime TIMESTAMP,
                dayOfWeek VARCHAR(10),
                plate VARCHAR(10),
                time_short TIME,
                prediction DOUBLE PRECISION
            )
        """).format(sql.Identifier(table_name))
        
        cursor.execute(create_table_query)
        conn.commit()
        print(f"Table '{table_name}' created")
        return True
    except psycopg2.Error as e:
        print(f"Error creating table: {e}")
        return False

def parse_time_short(arrival_time):
    """Parse arrival time to time_short format (HH:MM:SS)"""
    try:
        if 'T' in arrival_time:
            time_part = arrival_time.split('T')[1]
        else:
            time_part = arrival_time.split(' ')[1]
        
        parts = time_part.split(':')
        if len(parts) >= 2:
            hours = parts[0]
            minutes = parts[1]
            seconds = parts[2] if len(parts) > 2 else '00'
            return f"{hours}:{minutes}:{seconds}"
        return "00:00:00"
    except Exception as e:
        print(f"Error parsing time: {e}")
        return "00:00:00"

def calculate_time_duration(arrival_time, departure_time):
    """Calculate time difference in seconds"""
    try:
        # Replace 'Z' with '+00:00' to handle UTC timezone format explicitly
        arrival_dt = datetime.fromisoformat(arrival_time.replace('Z', '+00:00'))
        departure_dt = datetime.fromisoformat(departure_time.replace('Z', '+00:00'))
        return (departure_dt - arrival_dt).total_seconds()
    except Exception as e:
        print(f"Error calculating time duration: {e}")
        return 0

def load_json_data(file_path):
    """Load and sort JSON data"""
    try:
        with open(file_path, 'r') as f:
            data = json.load(f)
        
        if not isinstance(data, list):
            print("ERROR: JSON file must contain an array of objects")
            sys.exit(1)
        
        # Sort data by 'time' (duration) if available, or keep order. default 0 handles missing key.
        data.sort(key=lambda x: x.get('time', 0))
        return data
    except json.JSONDecodeError as e:
        print(f"ERROR in input file. It is not in valid JSON format: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"Error reading file: {e}")
        sys.exit(1)

def insert_data(conn, table_name, data):
    """Insert data into the table"""
    try:
        cursor = conn.cursor()
        
        insert_query = sql.SQL("""
            INSERT INTO {} (time, departureTime, arrivalTime, dayOfWeek, plate, time_short)
            VALUES (%s, %s, %s, %s, %s, %s)
        """).format(sql.Identifier(table_name))
        
        batch_data = []
        print("Inserting data into database...")
        
        for item in data:
            time_short = parse_time_short(item['arrivalTime'])
            # Calculate duration in seconds
            hours = calculate_time_duration(item['arrivalTime'], item['departureTime'])
            
            # Format timestamps for SQL compatibility (removing T separator)
            departure_time = item['departureTime'].replace('T', ' ')
            arrival_time = item['arrivalTime'].replace('T', ' ')
            
            batch_data.append((
                hours,
                departure_time,
                arrival_time,
                item['dayOfWeek'],
                item['plate'],
                time_short
            ))
        
        execute_batch(cursor, insert_query, batch_data)
        conn.commit()
        print(f"Successfully inserted {len(data)} records into '{table_name}'")
        return True
    except psycopg2.Error as e:
        print(f"Error inserting data: {e}")
        conn.rollback()
        return False

def main():
    # Check if input file exists
    if not os.path.exists(args.file_json):
        print(f"ERROR: Input file '{args.file_json}' does not exist")
        sys.exit(1)
    
    # 1. Connect to 'postgres' (default DB) to perform administrative tasks (CREATE DATABASE)
    # First connect to default database for database operations
    default_conn = create_connection("postgres", args.user, args.password, args.host, autocommit=True)
    
    if args.create:
        print("Creating database...")
        if not create_database(default_conn, args.database):
            print("ERROR: Failed to create database")
            sys.exit(1)
    
    # 2. Verify target database exists before proceeding
    # Check if database exists
    if not check_database_exists(default_conn, args.database):
        print(f"ERROR: Database '{args.database}' does not exist")
        default_conn.close()
        sys.exit(1)
    
    default_conn.close()
    
    # 3. Connect to the actual target database
    # Now connect to the target database
    conn = create_connection(args.database, args.user, args.password, args.host)
    
    # 4. Ensure the schema (table) exists
    # Check and create table if needed
    if not check_table_exists(conn, args.table):
        print("Creating table...")
        if not create_table(conn, args.table):
            print("ERROR: Failed to create table")
            conn.close()
            sys.exit(1)
    
    # 5. Load JSON and insert data into the table
    # Load and insert data
    data = load_json_data(args.file_json)
    if not insert_data(conn, args.table, data):
        print("ERROR: Failed to insert data")
        conn.close()
        sys.exit(1)
    
    conn.close()
    print("Data loading completed successfully")

if __name__ == "__main__":
    main()